sdk/include/reactos/shellutils.h

```diff
@@ -227,25 +227,25 @@ class CComDebugObject : public Base
 #endif
 
 template<class T>
-void ReleaseCComPtrExpectZero(CComPtr<T>& cptr, BOOL forceRelease = FALSE)
+ULONG ReleaseCComPtrExpectZeroHelper(CComPtr<T>& cptr, BOOL forceRelease = FALSE)
 {
+    ULONG r = 0;
     if (cptr.p != NULL)
     {
         T *raw = cptr.Detach();
-        int nrc = raw->Release();
-        if (nrc > 0)
+        int nrc = r = raw->Release();
+        while (nrc > 0 && forceRelease)
         {
-            DbgPrint("WARNING: Unexpected RefCount > 0 (%d)!\n", nrc);
-            if (forceRelease)
-            {
-                while (nrc > 0)
-                {
-                    nrc = raw->Release();
-                }
-            }
+            nrc = raw->Release();
         }
     }
+    return r;
 }
+#define ReleaseCComPtrExpectZero(...) \
+    if (ULONG r = ReleaseCComPtrExpectZeroHelper(__VA_ARGS__)) \
+        DbgPrint("WARNING: Unexpected RefCount > 0 (%d)!\n", r); \
+    else \
+        (void)0
```

Comment:
```suggestion
do {
    if ((ULONG r = ReleaseCComPtrExpectZeroHelper(__VA_ARGS__))) \
        DbgPrint("WARNING: Unexpected RefCount > 0 (%d)!\n", r); \
} while (0)
```

Reply:
And in which scenario does this make it work where the old code does not? Meaning, where can you use a `do` where the `if` will fail?

Reply:
https://stackoverflow.com/questions/154136/why-use-apparently-meaningless-do-while-and-if-else-statements-in-macros
(also answer for the other comment)

Reply:
I may be dense but that answers nothing. My code does not suffer dangling else. So again, in which scenario is the `do` better than the `if` we are looking at here?

Reply:
The general idea with the `do..while(0)` construct is that the macro has the same semantics as a normal statement.
E.g. no scope breaking stuff, ';' is required after it, and can be used in anything where you would expect to us a single statement.

Reply:
> The general idea with the `do..while(0)` construct is that the macro has the same semantics as a normal statement

And my claim is that `#define foo(x) if (x) { something(); } else (void)0` has the same semantics.

Reply:
It might have mostly the same semantics, but the `do..while(0)` thing is the most common way to do this,
and thus will be recognized by a lot more people for what it is.